loopy/transform/loop_fusion.py

```diff
@@ -0,0 +1,643 @@
+                continue
+
+            loops_to_be_fused.add(next_loop_in_queue)
+
+            for succ in ldg.successors[next_loop_in_queue]:
+                if ldg.is_infusible.get((next_loop_in_queue, succ), False):
+                    non_fusible_loops.add(succ)
+                else:
+                    queue.append(succ)
+
+        ldg = ldg.remove_nodes(loops_to_be_fused)
+        fused_chunks[name_gen(prefix)] = loops_to_be_fused
+
+    assert reduce(frozenset.union, fused_chunks.values(), frozenset()) == candidates
+    assert sum(len(val) for val in fused_chunks.values()) == len(candidates)
+
+    return fused_chunks
+
+
+def get_kennedy_unweighted_fusion_candidates(
+    kernel: LoopKernel, candidates: FrozenSet[str], prefix: str = "ifused"
+) -> Mapping[str, FrozenSet[str]]:
+    """
+    Returns the fusion candidates mapping that could be fed to
+    :func:`rename_inames_in_batch` similar to Ken Kennedy's Unweighted
+    Loop-Fusion Algorithm.
+
+    .. attribute:: prefix
+
+        Prefix for the fused inames.
+    """
+    from loopy.kernel.data import ConcurrentTag
+    from loopy.schedule.tools import (
+        _get_iname_to_tree_node_id_from_partial_loop_nest_tree,
+        get_partial_loop_nest_tree,
+    )
+
+    assert isinstance(kernel, LoopKernel)
+    assert isinstance(candidates, frozenset)
+
+    vng = kernel.get_var_name_generator()
+    fused_chunks = {}
+
+    # {{{ handle concurrent inames
+
+    # filter out concurrent loops.
+    all_concurrent_tags = reduce(
+        frozenset.union,
+        (kernel.inames[iname].tags_of_type(ConcurrentTag) for iname in candidates),
+        frozenset(),
+    )
+
+    concurrent_tag_to_inames = {tag: set() for tag in all_concurrent_tags}
+
+    for iname in candidates:
+        if kernel.inames[iname].tags_of_type(ConcurrentTag):
+            # since ConcurrentTag is a UniqueTag there must be exactly one of
+            # it.
+            (tag,) = kernel.tags_of_type(ConcurrentTag)
+            concurrent_tag_to_inames[tag].add(iname)
+
+    for inames in concurrent_tag_to_inames.values():
+        fused_chunks[vng(prefix)] = inames
+        candidates = candidates - inames
+
+    # }}}
+
+    tree = get_partial_loop_nest_tree(kernel)
+    iname_to_tree_node_id = _get_iname_to_tree_node_id_from_partial_loop_nest_tree(
+        tree
+    )
+
+    # {{{ sanitary checks
+
+    _nest_tree_id_to_candidate = {}
+
+    for iname in candidates:
+        loop_nest_tree_node_id = iname_to_tree_node_id[iname]
+        if loop_nest_tree_node_id not in _nest_tree_id_to_candidate:
+            _nest_tree_id_to_candidate[loop_nest_tree_node_id] = iname
+        else:
+            conflict_iname = _nest_tree_id_to_candidate[loop_nest_tree_node_id]
+            raise LoopyError(
+                f"'{iname}' and '{conflict_iname}' "
+                "cannot be fused as they can be nested "
+                "within one another."
+            )
```

Comment:
Suppose the CEESD version (see above) is the desired one, it's algorithmically wrong. If we presume the existence of a "fusable-with" relation, and if we also assume that that relation is transitive, then the correct thing to do upon discovering a non-fusable edge within the set of candidates would be to make two new sets of candidates. My understanding is that the net effect of the code there is to chuck the iname for which the conflict has been discovered out of the set and keep going, effectively (if done often enough) discarding one of the two candidate sets unnecessarily.

For specificity, [this](https://github.com/illinois-ceesd/loopy/blob/6b48f313955eb14023e65990215c72b6f4e72ebe/loopy/transform/loop_fusion.py#L766-L767) is the CEESD version I'm discussing.

Reply:
Yes, there's no need for the "is_infusible_with" relation and I got rid of it. See https://github.com/inducer/arraycontext/pull/217 for a correct usage of this transformation.

Reply:
Hi Matt!

> would you expect this recent change to have an effect on the issue

No I don't expect this change to have an issue on #453. I was just cleaning up the API here.


> it's an error now instead of a warning

The idea of that check is that if loops can be potentially nested within each other they should not be fusion candidates. Additionally, in our array context we pick candidates such that they don't violate this criterion.

We are hitting either one of the 2 issues:
1. They cannot be nested within each other and the check in the loop fusion implementation is buggy, or,
2. we are passing incorrect candidates from the FusionArrayContext.


If there's an MWE I'll be happy to take a look.

Reply:
Gotcha, thanks Kaushik! I think we are hitting issue #2 (so we'll just need to distinguish base and quadrature DOF axes).